firecracker.go

```diff
@@ -16,109 +16,239 @@ package firecracker
 import (
 	"context"
 	"github.com/go-openapi/strfmt"
+	"net"
+	"net/http"
 	"time"
 
 	"github.com/sirupsen/logrus"
 
 	"github.com/firecracker-microvm/firecracker-go-sdk/client"
 	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
 	ops "github.com/firecracker-microvm/firecracker-go-sdk/client/operations"
+	httptransport "github.com/go-openapi/runtime/client"
 )
 
 const firecrackerRequestTimeout = 500 * time.Millisecond
 
-// FirecrackerClient is a client for interacting with the Firecracker API
-type FirecrackerClient struct {
+// newFirecrackerClient creates a FirecrackerClient
+func newFirecrackerClient(socketPath string, logger *logrus.Entry, debug bool) *client.Firecracker {
+	httpClient := client.NewHTTPClient(strfmt.NewFormats())
+
+	socketTransport := &http.Transport{
+		DialContext: func(ctx context.Context, network, path string) (net.Conn, error) {
+			addr, err := net.ResolveUnixAddr("unix", socketPath)
+			if err != nil {
+				return nil, err
+			}
+
+			return net.DialUnix("unix", nil, addr)
+		},
+	}
+
+	transport := httptransport.New(client.DefaultHost, client.DefaultBasePath, client.DefaultSchemes)
+	transport.Transport = socketTransport
+
+	if debug {
+		transport.SetDebug(debug)
+	}
+
+	if logger != nil {
+		transport.SetLogger(logger)
+	}
+
+	httpClient.SetTransport(transport)
+
+	return httpClient
+}
+
+// ClientOpt is a functional option used to modify the client after construction.
+type ClientOpt func(*Client)
+
+// WithOpsClient will return a functional option and replace the operations
+// client. This is useful for mock and stub testing.
+func WithOpsClient(opsClient ops.ClientIface) ClientOpt {
+	return func(c *Client) {
+		c.client.Operations = opsClient
+	}
+}
+
+// Client is a client for interacting with the Firecracker API
+type Client struct {
 	client *client.Firecracker
 }
 
-// NewFirecrackerClient creates a FirecrackerClient
-func NewFirecrackerClient(socketPath string, logger *logrus.Entry, debug bool) *FirecrackerClient {
-	unixTransport := client.NewUnixSocketTransport(socketPath, logger, debug)
-	firecracker := client.New(unixTransport, strfmt.NewFormats())
+// NewClient creates a Client
+func NewClient(socketPath string, logger *logrus.Entry, debug bool, opts ...ClientOpt) *Client {
```

Comment:
This will require some thought due to how the code is currently laid out.

Reply:
So, after some thought, we probably should leave as is. The reason for this is we should make it incredibly easy to enable/disable logging and debugging. If we decide to go the other route and require functional options, it would need to be at the transport level. Meaning it would force users to create a custom transport with a custom API client then finally create a firecracker client. That seems like a lot of work to enable logging and debugging. What do you think?

Reply:
As for me the only parameter we should force user to pass - `socketPath` which is necessary for client to work properly. Everything else is nice to have customizations, but they are not required for correct functioning. So it's still good to able to configure debugging, to can be hidden to opts.